### PR TITLE
Add support for highlighting matching portions of a completion item when filtering.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Add support for highlighting matching portions of a completion item when filtering.

Also add support for completion item filtering.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/5589/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/5589/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/dbg/unit64 please

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/dbg/unit32 please

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/dbg/eta please

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
retest eta please

**Comment by CyrusNajmabadi** [comment]
Ping @dotnet/roslyn-ide

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs [hunk 1]
```diff
@@ -83,6 +103,19 @@ private CompletionItem GetCompletionItem(CompletionItem item)
...
                 }
 
+                var filterState = model.FilterState;
+
+                // If all the filters are on, or all the filters are off then we don't actually 
+                // need to filter.
+                if (filterState != null)
+                {
+                    if (filterState.Values.All(b => b) ||
+                        filterState.Values.All(b => !b))
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: cute

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs [hunk 2]
```diff
@@ -104,20 +137,33 @@ private CompletionItem GetCompletionItem(CompletionItem item)
                 CompletionItem bestFilterMatch = null;
                 bool filterTextIsPotentialIdentifier = false;
 
+                var itemToFilterText = new Dictionary<CompletionItem, string>();
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: Should we be pooling this?

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: I'm ok with it not being pooled.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionSet3.cs [hunk 3]
```diff
@@ -61,7 +95,7 @@ internal void SetTrackingSpan(ITrackingSpan trackingSpan)
                     var builderCompletion = ConvertCompletionItem(bestBuilder);
                     this.WritableCompletionBuilders.Add(builderCompletion);
 
-                    if (selectedItem.IsBuilder)
+                    if (selectedItem != null && selectedItem.IsBuilder)
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionSet3.cs]: ?.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionSet3.cs [hunk 4]
```diff
@@ -144,5 +177,60 @@ public override void Recalculate()
...
+#endif
+        {
+            if (_completionItemToFilterText != null)
+            {
+                var rules = this.GetCompletionRules();
+                if (rules != null)
+                {
+                    var completionItem = this._completionItemToVSCompletion.Keys.FirstOrDefault(k => k.DisplayText == displayText);
+
+                    if (completionItem != null && !completionItem.IsBuilder)
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionSet3.cs]: ?.

File: src/EditorFeatures/Core/Shared/Extensions/GlyphExtensions.cs [hunk 5]
```diff
@@ -317,8 +317,9 @@ public static ImageMoniker GetImageMoniker(this Glyph glyph)
                 case Glyph.Label:
                     return KnownMonikers.Label;
 
+                case Glyph.Parameter:
                 case Glyph.Local:
-                    return KnownMonikers.FieldPublic;
+                    return KnownMonikers.LocalVariable;
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Shared/Extensions/GlyphExtensions.cs]: FWIW, I thought this looked a **little** weird. Not a big deal though.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 6]
```diff
@@ -88,6 +85,58 @@ internal class SymbolCompletionItem : CompletionItem
...
+                // Don't allocate in the common case of just one symbol.
+                return GetFilters(symbols[0]);
+            }
+
+            var result = ImmutableArray<CompletionItemFilter>.Empty;
+            foreach (var symbol in symbols)
+            {
+                result = result.AddRange(GetFilters(symbol));
+            }
+            return result;
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: StyleCop sez... line break here :smile:

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 7]
```diff
@@ -88,6 +85,58 @@ internal class SymbolCompletionItem : CompletionItem
...
+                    }
+                    break;
+                case SymbolKind.RangeVariable:
+                case SymbolKind.Local:
+                case SymbolKind.Parameter: return CompletionItemFilter.LocalAndParameterFilters;
+            }
+
+            return ImmutableArray<CompletionItemFilter>.Empty;
+        }
+
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: return a match expression. Oh wait. Crap.

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 8]
```diff
@@ -913,4 +913,49 @@ This version used in: {2}</value>
...
+  </data>
+  <data name="Properties" xml:space="preserve">
+    <value>Properties</value>
+  </data>
+  <data name="Structures" xml:space="preserve">
+    <value>Structures</value>
+  </data>
+  <data name="Parameters" xml:space="preserve">
+    <value>Parameters:</value>
+  </data>
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/FeaturesResources.resx]: Kind of wish this was switched with Locals_and_parameters above

**Comment by DustinCampbell** [comment]
:+1:

**Pull request merged**